_posts/2017-08-16-Refactoring-with-ReactJS-testing.md

```diff
@@ -0,0 +1,253 @@
+{% highlight javascript linenos %}
+import { renderComponent, expect } from '../../test_helper.jsx';
+import FailureBanner from '../../../jsx/failureBanner.jsx';
+
+describe("failure banner", function(){
+    var component;
+
+    beforeEach(function(done){
+        component = renderComponent(FailureBanner, {message: 'test this failure message', isShown: true});
+        done();
+    })
+
+    it("renders failure banner", function(){
+        expect(component.$el.data('test-component')).to.equal('failureBanner');
+    })
+});
+{% endhighlight %}
+
+Let's go through the code play-by-play:
+{% highlight javascript %}
+  import { renderComponent, expect } from '../../test_helper.jsx';
+{% endhighlight %}
+Chai - the JS test assertion library uses function `expect`. We are importing `expect` from our helper simply because we have a wrapper for a regular Chai functionality that outputs a bit more of information if assertion fails.
+`renderComponent` is a generic component rendering wrapper that is suitable for any DOM rendering container. We will discuss it later as we are going through the helper code.
+
+{% highlight javascript %}
+  import FailureBanner from '../../../jsx/failureBanner.jsx';
+{% endhighlight %}
+Importing tested component from its relative path. Here as everywhere you might notice that we are importing components with their JSX extensions. This is a shortcoming in our build configuration that we were not yet able to overcome. Presently if you omit JSX it will get resolved to JS and fail to find the component.
+<p />Following `describe` method is comes from Mocha framework to wrap a suit of tests.<p />
+Let's look inside `beforeEach` call and dive into a bit more of a detail on rendering the component for testing.
+{% highlight javascript linenos %}
+beforeEach(function(done){
+        component = renderComponent(FailureBanner, {message: 'test this failure message', isShown: true});
+        done();
+    });
+{% endhighlight %}
+As we set out to render a component using our `renderComponent` helper method we have to pass in the ReactJS component being renderd as well as an object that reporesents properties we would like for this object to assume prior to rendering. As you can see the object is going to be shown with the included message.
+It is important to include `done` parameter into the function call and then to call it. We need to ensure that rendering of the object was completed before any tests have been performed on the rendered object. As you can see done parameter is a function that needs to be called to indicate that processing has been completed and the rest of the test can be executed.
```

Comment:
I'm not sure that this information i accurate. You shouldn't need to include `done` at all. `done` is meant to be used as a callback when you are writing asynchronous tests. There's nothing async going on here. You should able to just do 
```javascript
beforeEach(function(){
    component = renderComponent(FailureBanner, {message: 'test this failure message', isShown: true});
});
```
If that's not working, then I suspect there's some delay that karma is causing and it's a coincidence that the `done` callback is giving it just enough time to catch up. Let's sync up on that.

Reply:
Looks like tech debt to me. No good reason for it to be there.